File: java/tour-of-restate/src/main/java/dev/restate/tour/part1/TicketService.java
/*
 * Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Tour of Restate Java,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/tour-of-restate
 */

package dev.restate.tour.part1;

import com.google.protobuf.BoolValue;
import dev.restate.sdk.RestateContext;
import dev.restate.sdk.common.TerminalException;
import dev.restate.tour.generated.TicketServiceRestate;
import dev.restate.tour.generated.Tour.Ticket;

import java.time.Duration;

public class TicketService extends TicketServiceRestate.TicketServiceRestateImplBase {

    @Override
    public BoolValue reserve(RestateContext ctx, Ticket request) throws TerminalException {
        ctx.sleep(Duration.ofSeconds(35));
        return BoolValue.of(true);
    }

    @Override
    public void unreserve(RestateContext ctx, Ticket request) throws TerminalException {
    }

    @Override
    public void markAsSold(RestateContext ctx, Ticket request) throws TerminalException {
    }
}

File: typescript/tour-of-restate/src/app/ticket_service.ts
/*
 * Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Tour of Restate Typescript handler API,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/tour-of-restate
 */

import * as restate from "@restatedev/restate-sdk";

enum TicketStatus {
  Available,
  Reserved,
  Sold,
}

const reserve = async (ctx: restate.RpcContext) => {
  return true;
};

const unreserve = async (ctx: restate.RpcContext) => {
  return true;
};

const markAsSold = async (ctx: restate.RpcContext) => {
  return true;
};

export const ticketDbRouter = restate.keyedRouter({ reserve, unreserve, markAsSold });

export const ticketServiceApi: restate.ServiceApi<typeof ticketDbRouter> = {
  path: "TicketService",
};
